src/acl_kernel_if.cpp

```diff
@@ -470,7 +470,12 @@ static uintptr_t acl_kernel_cra_set_segment_rom(acl_kernel_if *kern,
 }
 
 static int acl_kernel_cra_read(acl_kernel_if *kern, unsigned int accel_id,
-                               unsigned int addr, unsigned int *val) {
+                               unsigned int addr, unsigned int *val,
+                               bool assert_csr_version = true) {
+  // Allow the very first reading csr_version call to run.
+  if (assert_csr_version) {
+    assert(kern->csr_version);
+  }
```

Comment:
Thanks @pcolberg. Based on my testing, trying to read/write if the CSR is absent will result the simulation hang. The underlying aocl_mmd_read() call just hangs.

I am trying to avoid the hang scenario by adding the assertion here, after setting the kern->csr_version.

Reply:
Thank @pcolberg . I store the `cra_ring_root_exist` in the `acl_kernel_if` now. To be safe, I will run SYCL-l3 for my changes.

Reply:
Thanks @pcolberg. Actually I also wanted to set the `kern->cra_ring_root_exist` in the `acl_kernel_if_update()`
 yesterday, at the beginning.
I didn't go with that approach because of the [acl_kernel_if_post_pll_config_init()](https://github.com/intel/fpga-runtime-for-opencl/blob/9a313e7ff2e93318464dc1bcd718a86f125835ef/src/acl_kernel_if.cpp#L1020) function, which actually does the CSR version checking, is called a couple of times **before** the `acl_kernel_if_update()`

I added some logs when trying to run a design in simulation.

```
 env CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1 ./test.exe
l_try_device is being called
acl_kernel_if_post_pll_config_init is being called 
l_try_device is being called
acl_kernel_if_post_pll_config_init is being called 
l_try_device is being called
acl_kernel_if_post_pll_config_init is being called 
l_try_device is being called
acl_kernel_if_post_pll_config_init is being called 
l_try_device is being called
acl_kernel_if_post_pll_config_init is being called 
l_try_device is being called
acl_kernel_if_post_pll_config_init is being called 
l_try_device is being called
acl_kernel_if_post_pll_config_init is being called 
acl_hal_mmd_program_device is being called
acl_hal_mmd_program_device is being called
**acl_kernel_if_update** is being called
kern->cra_ring_root_exist is 0
acl_kernel_if_post_pll_config_init is being called 
PASSED
```
`acl_kernel_if_post_pll_config_init ` is being called by `l_try_device` a couple of times, before the actual auto-discovery string is being read.
During the init() function, the simulatior was reading a predefined fake autodiscovery string here: [https://github.com/intel/fpga-runtime-for-opencl/blob/9a313e7ff2e93318464dc1bcd718a86f125835ef/src/acl_kernel_if.cpp#L716](https://github.com/intel/fpga-runtime-for-opencl/blob/9a313e7ff2e93318464dc1bcd718a86f125835ef/src/acl_kernel_if.cpp#L716)
 The simulation only got the real autodiscovery string when `acl_hal_mmd_program_device ` is being called.
 
Although it won't fail the test even if I only set the` kern->cra_ring_root_exist `only in the `acl_kernel_if_update` , because the the `kern->num_accel` was also 0 before the real autodiscovery string is read, so the `acl_kernel_cra_read` wasn't called. 

However, it might be a bit risky especially if there is no default value for kern->cra_ring_root_exist. Klocwork might complain about risky reading a uninitialized value

What do you think?

```
if (**kern->num_accel** > 0 && cra_ring_root_exist) {
    acl_kernel_cra_read(kern, 0, KERNEL_OFFSET_CSR, &csr);
    version = ACL_KERNEL_READ_BIT_RANGE(csr, KERNEL_CSR_LAST_VERSION_BIT,
                                        KERNEL_CSR_FIRST_VERSION_BIT);
    kern->csr_version = version;
    ACL_KERNEL_IF_DEBUG_MSG(kern,
                            "Read CSR version from kernel 0: Version = %u\n",
                            kern->csr_version);
  }```

Reply:
Thanks @pcolberg. I think it's a great idea to set the default value to be false! I have made the change accordingly, still working on the L3 testing.